06_puppy/Jat/puppyCRUD.go

```diff
@@ -0,0 +1,159 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	pupStore := MapStore{}
+	pupStore.Initialize()
+
+	pup1 := pupStore.Create("Beagle", "brown", "Sam")
+	pup2 := pupStore.Read(pup1.ID)
+	pup2.Value = "Bingo"
+	pupStore.Update(pup2)
+	pupStore.Read(1)
+	pup4 := pupStore.Create("Bull", "brown", "Tyson")
+	pup5 := pupStore.Read(pup4.ID)
+	pupStore.Delete(pup1)
+	pupStore.Delete(pup5)
+	pupStore.Create("rottweiler", "black", "Tipu")
+
+	pupSyncStore := SyncMapStore{}
+	pupSyncStore.Initialize()
+
+	pupSync1 := pupSyncStore.Create("Pom", "Black", "Sim")
+	pupSync2 := pupSyncStore.Read(pupSync1.ID)
+	pupSync2.Value = "Dingo"
+	pupSyncStore.Update(pupSync2)
+	pupSyncStore.Read(1)
+	pupSync4 := pupSyncStore.Create("Pug", "white", "Dim")
+	pupSync5 := pupSyncStore.Read(pupSync4.ID)
+	pupSyncStore.Delete(pupSync1)
+	pupSyncStore.Delete(pupSync5)
+	pupSyncStore.Create("Woulf", "Ash", "Togo")
+	fmt.Fprintln(out, fmt.Sprintf("Items remaining in Store %d", len(pupStore.m)))
+	pupStore3 := MapStore{}
+	initStore(&pupStore3)
+	accessInterface(&pupStore3)
+}
+func initStore(i Initializer) {
+	i.Initialize()
+}
+
+func accessInterface(s Storer) {
+	pup1 := s.Create("Beagle", "brown", "Sam")
+	pup2 := s.Read(pup1.ID)
+	pup2.Value = "Bingo"
+	s.Update(pup2)
+	s.Read(1)
+	pup4 := s.Create("Bull", "brown", "Tyson")
+	pup5 := s.Read(pup4.ID)
+	s.Delete(pup1)
+	s.Delete(pup5)
+	s.Create("rottweiler", "black", "Tipu")
+}
+
+type Puppy struct {
+	ID     int
+	Breed  string
+	Colour string
+	Value  string
+}
+type Initializer interface {
+	Initialize()
+}
+type Storer interface {
+	Create(string, string, string) Puppy
+	Read(int) Puppy
+	Update(Puppy) int
+	Delete(Puppy) int
+}
+
+//Map Store
+type MapStore struct {
+	m    map[int]Puppy
+	sqid int
+}
+
+//Initializer Implementation for Mapstore
+func (store *MapStore) Initialize() {
```

Comment:
A more go-idiomatic way would be to have constructor/factory type function prefixed with `New`:
```go
func NewMapStore() *MapStore {
    return &MapStore{m: map[int]Puppy{}}
}
```
but this is fine too.